samza-test/src/test/java/org/apache/samza/test/operator/StreamApplicationIntegrationTestHarness.java

```diff
@@ -0,0 +1,245 @@
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.protocol.SecurityProtocol;
+import org.apache.kafka.common.serialization.StringSerializer;
+import org.apache.samza.application.StreamApplication;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.KafkaConfig;
+import org.apache.samza.config.MapConfig;
+import org.apache.samza.runtime.ApplicationRunner;
+import org.apache.samza.test.harness.AbstractIntegrationTestHarness;
+import scala.Option$;
+
+import java.io.File;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * Harness for writing integration tests for {@link StreamApplication}s.
+ *
+ * <p> This provides the following features for its sub-classes:
+ * <ul>
+ *   <li>
+ *     Automatic Setup and teardown: Any non-trivial integration test brings up components like Zookeeper
+ *     servers, Kafka brokers, Kafka producers and Kafka consumers. This harness initializes each
+ *     of these components in {@link #setUp()} and shuts down each of them cleanly in {@link #tearDown()}.
+ *     {@link #setUp()} and {@link #tearDown()} are automatically invoked from the Junit runner.
+ *   <li>
+ *     Interaction with Kafka: The harness provides convenience methods to interact with Kafka brokers, consumers
+ *     and producers - for instance, methods to create topics, produce and consume messages.
+ *   <li>
+ *     Config defaults: Often Samza integration tests have to setup config boiler plate
+ *     to perform even simple tasks like producing and consuming messages. This harness provides default string
+ *     serdes for producing / consuming messages and a default system-alias named "kafka" that uses the
+ *     {@link org.apache.samza.system.kafka.KafkaSystemFactory}.
+ *   <li>
+ *     Debugging: At times, it is convenient to debug integration tests locally from an IDE. This harness
+ *     runs all its components (including Kafka brokers, Zookeeper servers and Samza) locally.
+ * </ul>
+ *
+ * <p> <i> Implementation Notes: </i> <br/>
+ * State persistence: {@link #tearDown()} clears all associated state (including topics and metadata) in Kafka and
+ * Zookeeper. Hence, the state is not durable across invocations of {@link #tearDown()} <br/>
+ *
+ * Execution model: {@link StreamApplication}s are run as their own {@link org.apache.samza.job.local.ThreadJob}s.
+ * Similarly, embedded Kafka servers and Zookeeper servers are run as their own threads.
+ * {@link #produceMessage(String, int, String, String)} and {@link #getMessages(Collection, int)} are blocking calls.
+ *
+ * <h3>Usage Example</h3>
+ * Here is an actual test that publishes a message into Kafka, runs an application, and verifies consumption
+ * from the output topic.
+ *
+ *  <pre> {@code
```

Comment:
Slightly prefer code examples in Java docs - as opposed to looking for a separate link, opening the file through an IDE and viewing actual code. Also, Makes documentation clearer when viewed from web-pages and documentation tools. 

Agreed, that keeping documentation consistent with refactors is a minor tax to pay.